File: backend/src/main/java/com/example/backend/model/Promotion.java
package com.example.backend.model;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "promotions")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Promotion {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "promo_id")
    private UUID id;

    @Column(name = "code", nullable = false, length = 50, unique = true)
    private String code;

    @Column(name = "discount_percent", nullable = false)
    private Double discountPercent;

    @Column(name = "valid_from", nullable = false)
    private LocalDateTime validFrom;

    @Column(name = "valid_to", nullable = false)
    private LocalDateTime validTo;

    @Column(name = "max_uses")
    private Integer maxUses;

    @Builder.Default
    @Column(name = "used_count")
    private Integer usedCount = 0;

    @Column(name = "min_booking_amount")
    private Double minBookingAmount;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "promotion")
    private List<Booking> bookings;
}

File: frontend/src/types/promotion.ts
import type { BaseEntity, BaseFilterRequest } from "./common";

export interface PromotionResponse extends BaseEntity {
  code: string;
  discountPercent: number;
  validFrom: string;
  validTo: string;
  maxUses: number;
  usedCount: number;
  minBookingAmount: number;
}

export interface PromotionRequest {
  code: string;
  discountPercent: number;
  validFrom: string;
  validTo: string;
  maxUses: number;
  minBookingAmount: number;
}

export interface PromotionFilterRequest extends BaseFilterRequest {
  code?: string;
  validFrom?: string;
  validTo?: string;
  minBookingAmount?: number;
}